src/runtime/path_resolver.ts

```diff
@@ -0,0 +1,172 @@
+            // Make sure it is not a child of any prefixed path
+            for (const ancestor of PathResolver.pathAncestors(path.path)) {
+                const conflictingPrefix = takenPrefixes.get(ancestor);
+                if (conflictingPrefix) {
+                    issuePaths.add(conflictingPrefix);
+                    issuePaths.add(routeInfo);
+                }
+            }
+            
+            // If the path is a prefix:
+            if (path.isPrefix) {
+                // Make sure it is not an ancestor of any other registered path
+                // (prefix or exact)
+                const conflictingChild = bannedPrefixes.get(path.path);
+                if (conflictingChild) {
+                    issuePaths.add(conflictingChild);
+                    issuePaths.add(routeInfo);
+                }
+
+                takenPrefixes.set(path.path, routeInfo);
+            }
+
+            // Mark all ancestors of the path (and the path itself) as banned prefixes
+            for (const ancestor of PathResolver.pathAncestors(path.path)) {
+                bannedPrefixes.set(ancestor, routeInfo);
+            }
+
+            // Disallow exact matches for both prefixe and exact
+            takenExact.set(path.path, routeInfo);
+        }
+
+        // if (issuePaths.size > 0) {
+        //     throw new RouteCollisionError(Array.from(issuePaths));
+        // }
+        return issuePaths;
+    }
+
+    /**
+     * A function to look up `QualifiedPathPair` values by their path.
+     * 
+     * @param path The pathname the request was made to
+     * @returns The `T` value associated with the path, whether that is an exact
+     * or a prefixed path
+     */
+    public resolve(path: string): QualifiedPathPair | null {
+        // Try to find an exact match first (generally just an optimization)
+        if (this.exactRoutes.has(path)) {
+            return this.exactRoutes.get(path)!;
+        }
+
+        // Iterate over all ancestors of the path and check if any of the
+        // ancestors are a registered prefix.
+        //
+        // This is speedy enough because the runtime is tied to the length of
+        // the path, rather than the number of routes.
+        for (const prefix of PathResolver.pathAncestors(path)) {
+            if (this.prefixRoutes.has(prefix)) {
+                return this.prefixRoutes.get(prefix)!;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * A generator utility function that yields all ancestors of a given path.
+     * 
+     * This is used in collision checking to ensure that no path is a child of a
+     * prefix and no prefix is the ancestor of any path.
+     * 
+     * This is also used in path resolution to check for prefix matches.
+     * 
+     * @param path The URL pathname to generate all ancestors for
+     * 
+     * @yields The path itself and all ancestors, up to and including the root
+     */
+    public static *pathAncestors(path: string) {
```

Comment:
nit: should be standalone function since it provides generic functionality not specific to this object